### DISCUSSION TITLE
The sqlalchemy ORM doesn't work properly with IN clause.

### DISCUSSION THREAD

**Discussion by tkacikdominik** [sqlalchemy/sqlalchemy#11667] [Usage Questions] [answered]
### Describe the bug

The sqlalchemy ORM doesn't work properly with IN clause. My query is as follows
```python
with db_engine.connect() as conn:
    with Session(conn) as session:
        analysis_query = (
            select(ChemAnalysisDb)
            .where(extract("year", ChemAnalysisDb.analyse_date) == 2023)
            .where(ChemAnalysisDb.measurement.in_([12720, 22561, 12721, 12722]))
        )
       analyses = session.scalars(analysis_query).all()
```
Results of this query is two objects copied to the length of the real result.
```
[<....db_datamodel.ChemAnalysisDb at 0x23e20dbb550>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20a20d10>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20dbb550>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20a20d10>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20dbb550>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20a20d10>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20dbb550>,
 <....db_datamodel.ChemAnalysisDb at 0x23e20a20d10>]
```
I'm checked the compiled query and seems be OK. If I execute the query outside the ORM, result is OK.

```python
# doesn't work
analyses = session.scalars(analysis_query).all()
# work correctly
analyses = connection.execute(analysis_query)
```
The result is different objects with different data.
```
(1, 12720, datetime.datetime(2023, 11, 3, 17, 52), 4.39, 0.44, 0.43, 0.042, 0.038, 0.001, 0.001, 0.001, 0.014)
(2, 12720, datetime.datetime(2023, 11, 3, 18, 22), 4.39, 0.43, 0.42, 0.04, 0.034, 0.001, 0.001, 0.001, 0.014)
(1, 12721, datetime.datetime(2023, 11, 3, 19, 52), 4.47, 0.45, 0.35, 0.042, 0.04, 0.001, 0.001, 0.001, 0.012)
(2, 12721, datetime.datetime(2023, 11, 3, 20, 24), 4.43, 0.45, 0.35, 0.043, 0.037, 0.001, 0.001, 0.001, 0.012)
(1, 12722, datetime.datetime(2023, 11, 3, 22, 33), 4.25, 0.42, 0.37, 0.042, 0.047, 0.001, 0.001, 0.001, 0.011)
(2, 12722, datetime.datetime(2023, 11, 3, 23, 24), 4.25, 0.44, 0.44, 0.045, 0.036, 0.001, 0.001, 0.001, 0.012)
(1, 22561, datetime.datetime(2023, 11, 3, 22, 57), 4.46, 0.21, 0.49, 0.034, 0.021, 0.001, 0.001, 0.001, 0.022)
(2, 22561, datetime.datetime(2023, 11, 3, 23, 6), 4.71, 0.22, 0.52, 0.031, 0.017, 0.001, 0.001, 0.001, 0.024)
```

### SQLAlchemy Version in Use

2.0.30

### DBAPI (i.e. the database driver)

oracle+cx_oracle

### Database Vendor and Major Version

Oracle

### Python Version

3.11.4

### Operating system

Windows 10

### To Reproduce

```python
with db_engine.connect() as conn:
    with Session(conn) as session:
        analysis_query = (
            select(ChemAnalysisDb)
            .where(extract("year", ChemAnalysisDb.analyse_date) == 2023)
            .where(ChemAnalysisDb.measurement.in_([12720, 22561, 12721, 12722]))
        )
       # Bad result
       analyses = session.scalars(analysis_query).all()
       # OK result
       analyses2 = connection.execute(analysis_query)
```

**Comment by zzzeek** [comment, marked as answer]
this looks like an issue with your models.   [The ORM will return the same object for each occurrence of a particular primary key](https://docs.sqlalchemy.org/en/20/glossary.html#term-identity-map).  Your results look like your database table includes columns as part of the primary key that you didnt state in your model - note the two objects correspond to the first column alternating as "1" and "2" over and over again.   It looks like the column following it should also be part of the primary key, meaning you'd include `primary_key=True` in that column definition.

**Reply by tkacikdominik** [reply]
Thank you very much for the quick answer.

It's really a problem with the primary key column. I checked the table definition, and the table doesn't have any primary key and I also don't have permission to alter the table. Also, table doesn't have any unique column and it looks like that's the reason ORM is not usable for my usecase.

**Reply by zzzeek** [reply]
this can be accommodated, see https://docs.sqlalchemy.org/en/20/faq/ormconfiguration.html#how-do-i-map-a-table-that-has-no-primary-key
